### PR TITLE
api: make .open() stream data for Azure

### PR DESCRIPTION
**Pull request by Suor**
This employs generating temporary signed public url and then streaming
from it. The bulk of the code might be reused for other remotes.

Url streaming does resume on connection error using "Range" header and
regenerates public url too.

If remote class doesn't provide .open() then this falls back to pulling
a file and opening it locally.

### CODE REVIEW

**Comment by Suor** [comment]
So Azure is complete and tested. Windows tests fail for some other reason.

I would appreciate this to be merged. I will add support for other remotes separately.

**Comment by efiop** [comment]
@Suor Please take a look at codeclimate and deepsource repots.

**Comment by efiop** [comment]
Also, if you are no longer working on the patch, please remove WIP.

**Comment by Suor** [comment]
@efiop looked at them DeepSource issues are either not related, opinionated or broken. Fixing CodeClimate one will make it worse.

File: dvc/utils/http.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        def read1(self, n=-1):
+            try:
+                chunk = self.leftover or next(self.iterator)
+            except StopIteration:
+                return b""
+
+            if n > 0:
+                output, self.leftover = chunk[:n], chunk[n:]
+                return output
+            else:
```
Comment (efiop) [inline dvc/utils/http.py]: As DS noted, else is not needed.

File: dvc/utils/http.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        if mode == "rb":
+            yield bytes_stream
+        else:
+            encoding = encoding or response.encoding
+            yield io.TextIOWrapper(bytes_stream, encoding=encoding)
+
+
+@contextmanager
+def iter_url(url, chunk_size=io.DEFAULT_BUFFER_SIZE):
```
Comment (efiop) [inline dvc/utils/http.py]: Codeclimate is not happy about you cramping full blown functions into another function. If helpers are not trivial , consider moving them outside.

File: dvc/utils/http.py (hunk 1 above)
Comment (Suor) [reply dvc/utils/http.py]: I think it's more obvious this way.

File: dvc/utils/http.py (hunk 2 above)
Comment (Suor) [reply dvc/utils/http.py]: They don't make sense outside.

Comment (Suor) [reply dvc/utils/http.py]: BTW, code climate should add up complexity of nested functions, this is the same as summing up complexities of all the methods in a class )

File: dvc/utils/http.py (hunk 1 above)
Comment (efiop) [reply dvc/utils/http.py]: It is not and we don't do such things in other places. Please adjust.

File: dvc/utils/http.py (hunk 2 above)
Comment (efiop) [reply dvc/utils/http.py]: @Suor Class is not the same as a function, please split it. Again, we don't do such things in other places.

Comment (efiop) [reply dvc/utils/http.py]: Ok, we've discussed this privately and decided to go with this :slightly_smiling_face:

**Comment by efiop** [comment]
@Suor Patch style checks are failing. Please take a look.

File: setup.py [hunk 3]
```diff
@@ -107,6 +107,7 @@ def run(self):
     "jaraco.windows==3.9.2",
     "mock-ssh-server>=0.5.0",
     "moto>=1.3.14.dev55",
+    "rangehttpserver==1.2.0",
```
Comment (shcheklein) [inline setup.py]: @efiop @Suor this is not Apache 2 compatible I believe.

Also, I wonder what's happening with Conda. We'll need to wriate a checklist on adding new dependencies.

Comment (shcheklein) [reply setup.py]: sorry, false alarm ... there are two of them on Github

Comment (efiop) [reply setup.py]: @shcheklein I'm not sure our conda packages are running tests though. We've added this package only as a [tests] dependency.

**Pull request merged**